### ISSUE TITLE
Poor results on objaverse object

### ISSUE THREAD

**Issue by vlongle** [vuer-ai/feature-splatting-inria#7]
Hi,

Congratulations on your great work!

I'm trying to use your feature-splatting on a single Objaverse object. The RGB reconstruction looks reasonable but the decoded features look poor. Here's the code for visualizing various outputs
```
from scene.gaussian_model import GaussianModel
from scene import Scene
from gaussian_renderer import render
from arguments import ModelParams, PipelineParams, get_combined_args
from utils.general_utils import safe_state
import torch
import numpy as np
from PIL import Image
from argparse import ArgumentParser
import os
import torchvision
from sklearn.decomposition import PCA
import torch.nn.functional as F

# Set up command line argument parser
parser = ArgumentParser(description="Debug CLIP features from rendered Gaussians")
model = ModelParams(parser)  # Remove sentinel=True to properly set up params
pipeline = PipelineParams(parser)
parser.add_argument("--iteration", default=-1, type=int)
parser.add_argument("--quiet", action="store_true")
parser.add_argument("--camera_index", default=1, type=int)
args = get_combined_args(parser)

print("Rendering " + args.model_path)

# Initialize system state (RNG)
safe_state(args.quiet)

# Extract model and pipeline params
model_params = model.extract(args)
pipe = pipeline.extract(args)

# Initialize model
gaussians = GaussianModel(model_params.sh_degree, model_params.distill_feature_dim)
scene = Scene(model_params, gaussians, load_iteration=args.iteration, shuffle=False)

# Get specified camera view
camera_view = scene.getTrainCameras()[args.camera_index]

# Set up rendering parameters
bg_color = [1,1,1] if model_params.white_background else [0, 0, 0]
print("model_params.white_background?", model_params.white_background)
background = torch.tensor(bg_color, dtype=torch.float32, device="cuda")

# Render with features
render_output = render(camera_view, gaussians, pipe, background, render_features=True)

# Get results
rendered_image = render_output["render"] ## (3, H, W)
raw_rendered_features = render_output["render_feat"] ## (16, H, W)


def visualize_features(features, out_path):
    C, H, W = features.shape
    feature_nc = features.reshape((C, -1)).transpose((1, 0))

    # Replace any NaN values with 0
    feature_nc[np.isnan(feature_nc)] = 0

    # Fit PCA and transform the features to 3 dimensions
    pca = PCA(n_components=3)
    feature_3c = pca.fit_transform(feature_nc)

    # Scale the features to 0-255 range using percentile-based normalization
    quan_min = np.quantile(feature_3c, 0.05)
    quan_max = np.quantile(feature_3c, 0.95)
    feature_3c = (feature_3c - quan_min) / (quan_max - quan_min) * 255
    feature_3c = np.clip(feature_3c, 0, 255).astype(np.uint8)

    # Reshape back to image dimensions (H,W,3)
    feature_rgb = feature_3c.reshape((H, W, 3))

    # Create and display the image
    img = Image.fromarray(feature_rgb)
    img.save(out_path)


output_dir = "test"
os.makedirs(output_dir, exist_ok=True)
torchvision.utils.save_image(rendered_image, os.path.join(output_dir, "rendered_image.png"))
visualize_features(raw_rendered_features.detach().cpu().numpy(), os.path.join(output_dir, "rendered_features.png"))



from scene import skip_feat_decoder
my_feat_decoder = skip_feat_decoder(gaussians._distill_features.shape[1], part_level=False).cuda()
decoder_weight_path = os.path.join(args.model_path, "feat_decoder.pth")
assert os.path.exists(decoder_weight_path)
decoder_weight_dict = torch.load(decoder_weight_path)
my_feat_decoder.load_state_dict(decoder_weight_dict, strict=True)
my_feat_decoder.eval()

decoded_feat = my_feat_decoder(raw_rendered_features.unsqueeze(0))[1].squeeze(0)  # 768 x H x W
print("decoded_feat.shape", decoded_feat.shape) ## (768, 512, 512)
visualize_features(decoded_feat.detach().cpu().numpy(), os.path.join(output_dir, "decoded_features.png"))




gt_clip_feat = camera_view.feat_chw.cuda()
print("gt_clip_feat.shape", gt_clip_feat.shape) ## (768, 64, 64)

rendered_feat_bhwc = F.interpolate(raw_rendered_features.unsqueeze(0), size=gt_clip_feat.shape[1:], mode="bilinear", align_corners=False)
print("rendered_feat_bhwc.shape", rendered_feat_bhwc.shape) ## (16, 64, 64)
dino_feat, downsampled_clip_feat = my_feat_decoder(rendered_feat_bhwc)
print("downsampled_clip_feat.shape", downsampled_clip_feat.shape) ## (768, 64, 64)

downsampled_clip_feat = downsampled_clip_feat.squeeze(0)

visualize_features(downsampled_clip_feat.detach().cpu().numpy(), os.path.join(output_dir, "downsampled_clip_feat.png"))

visualize_features(gt_clip_feat.detach().cpu().numpy(), os.path.join(output_dir, "gt_clip_feat.png"))





## compute clip loss

from utils.loss_utils import l1_loss, ssim, l2_loss, cosine_loss

### TODO: NOTE: some weird ignoring masks going on here.
ignore_feat_mask = (torch.sum(gt_clip_feat == 0, dim=0) == gt_clip_feat.shape[0])
gt_clip_feat[:, ignore_feat_mask] = downsampled_clip_feat.squeeze(0)[:, ignore_feat_mask]

clip_loss = cosine_loss(downsampled_clip_feat, gt_clip_feat)
print("clip_loss", clip_loss)
```
We use 100 images rendered from BlenderNerf. The rendered RBG looks good, 

![Image](https://github.com/user-attachments/assets/2e69a6d0-eaaf-44db-9458-daf25313296e)

But the rendered features (16D in the latent space) visualized by PCA on 3 components

![Image](https://github.com/user-attachments/assets/d2a40553-8547-437c-b378-37ee59c59c68)

and decoded CLIP features (put thru the decoder)

![Image](https://github.com/user-attachments/assets/30a34e12-7907-4f41-90f5-092978627b55)

look poor. The gt_clip_feature is what we would expect to see

![Image](https://github.com/user-attachments/assets/0e2493c8-02fa-4229-a627-a127e4dd7c43)

where we can see the ficus, the pot and the background.

Do you have any idea why this might be happening or if there's something to try to fix this?

**Comment by RogerQi** [comment]
Thank you for the detailed explanation and the figures! My guess is that the culprit is floating invisible gaussian. For synthetic objects  like the ones in objaverse, since they have empty background, there would be Gaussians with low opacity / dark colors that don't get penalized during training.

To verify this, you can pull out the means of Gaussians, visualize them as point clouds in 3rd party software like open3d. If there are many unexpected floating point, then this is likely the reason.

Some potential fixes:
- During training, use opacity mask / randomized BG color to penalize floating Gaussians
- After training, remove Gaussians with low opacity and/or dark colors

I ran some experiments on objaverse as well, but I couldn't find the data processing scripts now... If you can share a script to convert objaverse assets into trainable format for me to reproduce this issue on my machine, I can find time to pinpoint this issue and push a solution.

**Comment by vlongle** [comment]
Thanks for the clarification. 

Here's the script for generating data using blenderNerf add-on.

[generate_blendernerf_data.txt](https://github.com/user-attachments/files/19243650/generate_blendernerf_data.txt)

which can be run as 
```
blender --background --python generate_blendernerf_data.py -- --object_path {your_data_path}/ecb91f433f144a7798724890f0528b23.glb --num_images 100 --format NERF --camera_dist 1.8
```

I used `ecb91f433f144a7798724890f0528b23` from objaverse

```
import objaverse


uid=["ecb91f433f144a7798724890f0528b23"]
objects = objaverse.load_objects(
uids=uid,
)
print("objects", objects)
```

**Comment by RogerQi** [comment]
Thanks for your prompt response. I found that the objaverse conversion script in my repo is still working. I took the uid of your assets and I got pretty good results. I think the root issue is how your code handles the alpha channel for the background.

I wrote a quick draft of command I use [here](https://github.com/vuer-ai/feature-splatting-inria/blob/roger/objaverse_feature/docs/synthetic_data_generation.md). I will try to find time in the next few days to consolidate all changes and update the README of the repo.

![Image](https://github.com/user-attachments/assets/6a5764c4-96a9-4608-9533-affb7ba20dc9)

![Image](https://github.com/user-attachments/assets/23101dc8-a112-4feb-9a84-d4630c425d06)

**Comment by RogerQi** [comment]
@vlongle I just ran a full-scale 30K iter training using the command on the ficus data. Since the camera sampling in my script is randomized and I don't have your exact camera viewport, I made interpolation video from top to bottom of the plant. Here are the results.

The commands I used for training is updated at: https://github.com/vuer-ai/feature-splatting-inria/blob/roger/objaverse_feature/docs/synthetic_data_generation.md

The commands I used for rendering the heatmap after training: python render.py -m output/ficus -s feat_data/objaverse/Ficus  --camera_slerp_list 13 31 --step_size 60 --with_feat --clip_feat --text_query "pot" --neg_text_query "plant, green leaf"

https://github.com/user-attachments/assets/e988e8f2-d9c6-4b2b-8916-5f1f86a4da93
https://github.com/user-attachments/assets/debec5e5-3c76-4955-9a09-56763c1b57f8

**Comment by vlongle** [comment]
Thanks for the update.

**Comment by vlongle** [comment]
I also just noticed that here we have to use very specific `neg_text_query` like "plant, green leaf" whereas some other nerf alternatives like LERF can detect fine-grained parts using generic negative text query like "objects,things,stuff,texture"

![Image](https://github.com/user-attachments/assets/3c5a6a68-4f22-40d6-926b-6fdd01158e62)

You said in the paper that Nerf learns continuous function so there's implicit regularization, but I'm wondering if the current feature splatting approach can still handle this.

**Comment by RogerQi** [comment]
I think it has to do with the synthetic data. Currently the heatmap is
generated by normalizing the clip scores. Since the synthetic image has no
background, the lower end of the scores are higher. If you check out the
nerfstudio example video I have on the website, I didn’t use specific
negative queries there.

**Comment by RogerQi** [comment]
I will close this issue now as I haven't received an update. Some final notes/ideas if you are looking to improve the quality of the features

- The ECCV paper uses full version of SAM, whereas this codebase uses MobileSAM for efficiency. If you are looking for part-level details, switching back to SAM may help
- Same reason for efficiency, we didn't compute multi-scale CLIP features. That also helps.

Feel free to send me an email or something if you want to discuss more.
